Prefix frontend dispatch

eddyashton opened the pull request:
This is an extension of #422, created as a separate branch and PR as it includes additional refactoring.

It aims to separate the parsing of transport-level framing information from the semantic transaction processing. In particular extracting an actor from the transaction message data, rather than from the certificate which established the underlying channel.

In our current framed JSON-RPC protocol the actor is prefixed on the method, so `"method": "users/getCommit"` will be dispatched by the endpoint to the `users` frontend, and by that frontend to its `getCommit` method handler. This requires a full parse of the JSON-RPC object before we can do the `valid_caller` check - in a flatbuffer or HTTP endpoint it should be possible to do a partial parse (of only the header) to extract the actor and similar framing information, and pass the remainder unprocessed data verbatim to the frontend.

The key code change here is that `RpcFrontend::process` now takes a single context object, populated with parsed details. This is largely to make API changes easier, so we can tweak this struct's contents (to narrow down what the frontend requires) without altering hundreds of calling points. This context is split into a common `SessionContext` which may persist over many transactions, and the remaining fields populated per-RPC. The unit tests call `process` rather than `process_json`, and have a gestures towards standardisation.

For future PRs:
- Cert changes
- Consolidation of `process_forwarded` and `process_pbft`
- Reduction of forwarded special-casing (how much framing should be re-parsed/re-validated on executor?)

eddyashton commented on src/enclave/rpchandler.h:
```diff
@@ -23,8 +23,8 @@ namespace enclave
     // Used by rpcendpoint to process incoming client RPCs
-    virtual std::vector<uint8_t> process(
-      RPCContext& ctx, const std::vector<uint8_t>& input) = 0;
+    virtual std::optional<std::vector<uint8_t>> process(
```
Rather than setting `ctx.is_pending`, this now indicates a pending transaction by returning `nullopt`. This lets us reason about the `RPCContext` as pure input (const).

cimetrics commented:
![images](https://raw.githubusercontent.com/microsoft/CCF/cimetrics/cimetrics/image2019-11-01%2014%3A23%3A55.270251.png)

eddyashton commented on src/node/rpc/forwarder.h:
```diff
@@ -63,22 +64,23 @@ namespace ccf
         serialized::write(data_, size_, caller_cert.data(), caller_cert.size());
       }
-      serialized::write(data_, size_, data.data(), data.size());
+      serialized::write(data_, size_, rpc_ctx.raw.data(), rpc_ctx.raw.size());
```
This feels like a half-finished bodge that works by chance. I _think_ we should either be sending everything in the context _apart_ from the original raw data, or _only_ the raw data.

achamayou replied:
Please ticket that.

jumaffre replied:
We'll have to keep forwarding `caller_id`, `rpc_ctx.session.client_session_id` and, optionally, `caller_cert` as these are not part of the raw data?

eddyashton replied:
Will deal with this in #511 when we can look at it in isolation.

achamayou commented on src/enclave/enclavetypes.h:
```diff
@@ -11,69 +11,127 @@ namespace enclave
+    SessionContext session;
+
+    // Packing format of original request, should be used to pack response
+    std::optional<jsonrpc::Pack> pack = std::nullopt;
```
We have a default one of those (text I think), for when we can't establish the packing format of the input. I think it should be the value here?

eddyashton replied:
Good point. We were setting this default on line 61 of `rpcendpoint.h`, but we might as well just do it more generally here.

achamayou commented on src/enclave/enclavetypes.h:
```diff
@@ -11,69 +11,127 @@ namespace enclave
+      rpc_ctx.unpacked_rpc = rpc.at(jsonrpc::REQ);
+      ccf::SignedReq signed_req;
+      signed_req.sig = sig_it->get<decltype(signed_req.sig)>();
+      signed_req.req = nlohmann::json::to_msgpack(rpc_ctx.unpacked_rpc);
```
It occurs to me that once we stop this madness, the client sig tx perf test will surely get better.

achamayou commented on src/enclave/rpcendpoint.h:
```diff
@@ -34,39 +35,93 @@ namespace enclave
-        caller = peer_cert();
+      auto search = rpc_map->find(actor);
+      if (!search.has_value())
+      {
```
I think we should at least LOG_DEBUG here?

eddyashton replied:
Sure - I've reinstated a `LOG_DEBUG` similar to the one we had previously.

jumaffre commented on src/enclave/enclavetypes.h:
```diff
@@ -11,69 +11,127 @@ namespace enclave
+    {
+      rpc_ctx.unpacked_rpc = rpc.at(jsonrpc::REQ);
+      ccf::SignedReq signed_req;
+      signed_req.sig = sig_it->get<decltype(signed_req.sig)>();
```
Why not using the original code in `clientsignatures.h`?

eddyashton replied:
Good catch, this is the same so simple assignment to call `from_json` should work.

jumaffre commented on src/enclave/enclavetypes.h:
```diff
@@ -11,69 +11,127 @@ namespace enclave
+    }
+
+    const auto method_it = rpc_ctx.unpacked_rpc.find(jsonrpc::METHOD);
+    if (method_it != rpc_ctx.unpacked_rpc.end())
```
Do we still need this bit?

eddyashton replied:
Yes - this is extracting the method (&co) out of the JSON-RPC, and everything after this should ideally only look at `ctx.method`. I'll do another pass in a future PR actually enforcing this if it works, if no-one else needs `ctx.unpacked_rpc` or `ctx.raw`.

jumaffre commented on src/node/rpc/frontend.h:
```diff
@@ -791,32 +708,18 @@ namespace ccf
-      if (!rpc.first)
+      // Store client signature. It is assumed that the forwarder node has
+      // already verified the client signature.
+      if (ctx.signed_request.has_value())
```
I think we need to wrap this with `if (consensus == nullptr || consensus->is_primary())` as, even if the command is forwarded, the primary may have changed in the meantime and the forwarded command should not write to the KV. It probably won't matter as no transaction will be committed but we never know.

eddyashton replied:
I think in that case we have to rely on the transaction failing later with an "I'm not primary" error? If we somehow _do_ execute this transaction and make changes to the kv, we should record the signature as well. More checks means more possible paths in buggy scenarios.

jumaffre commented on src/node/nodestate.h:
```diff
@@ -282,11 +282,17 @@ namespace ccf
       ctx.is_create_request = true;
 
-      frontend->process(ctx, packed);
+#ifdef PBFT
```
Why this? The request should go via `process()` no matter the consensus.

eddyashton replied:
Yeah this was based on an old git branch, and it looks like a merge reintroduced this `ifdef` branch. Removed.

jumaffre commented:
Sorry I have to be that guy: we need to update the docs accordingly. 
At least: https://microsoft.github.io/CCF/developers/example.html and https://microsoft.github.io/CCF/users/issue_commands.html

And also update the RPC API spec?

cimetrics commented:
![images](https://raw.githubusercontent.com/microsoft/CCF/cimetrics/cimetrics/image2019-11-01%2018%3A07%3A09.755605.png)

cimetrics commented:
![images](https://raw.githubusercontent.com/microsoft/CCF/cimetrics/cimetrics/image2019-11-04%2010%3A46%3A24.559665.png)

olgavrou commented on src/enclave/enclavetypes.h:
```diff
@@ -11,69 +11,124 @@ namespace enclave
+      throw std::logic_error(fmt::format("Failed to unpack: {}", rpc.dump()));
+    }
+
+    parse_rpc_context(rpc_ctx, rpc);
```
What about first un-packing the RPC, then calling `parse_rpc_context` which can create and return a new `RPCContext` object instead of having a `RPCContext` passed into the method and modifying it, and then setting `raw` and `SessionContext` to the returned `RPCContext`?

olgavrou commented on src/enclave/rpcendpoint.h:
```diff
@@ -34,39 +35,98 @@ namespace enclave
+      }
+      LOG_TRACE_FMT("Deserialised");
+
+      parse_rpc_context(rpc_ctx, rpc);
```
Seeing as you are doing the same here, what about unpacking first, and if everything is OK, passing `SessionContext`, `data` and `rpc` into `parse_rpc_context` and have that create and return your `RPCContext` object. That way you don't have to repeat this code.

eddyashton replied:
I'd prefer to keep the current format, as in general the context parsing is multi-stage - populating the context piece-by-piece as we peel off additional layers of wrapping. The JSON-RPC case doesn't show this very well since we only go through 3 stages (empty context tied to a session, detecting the packing format during `unpack_rpc`, extracting all other fields in `parse_rpc_context`), but its useful to be able to return appropriate errors between each stage. I want to keep this pattern, and if anything pull more stages out (extracting the signature perhaps), rather than storing features temporarily to build the finished context.

This causes some code repetition, but I think that's inevitable to support our tests. `handle_data` is the real parsing function, returning detailed error messages to a client if the parse fails. `make_rpc_context` does something similar for the tests, so they can call `process` directly.

olgavrou commented on tests/infra/jsonrpc.py:
```diff
@@ -282,15 +283,16 @@ def __init__(
 
-    def request(self, *args, **kwargs):
-        r = self.stream.request(*args, **kwargs)
+    def request(self, method, params, *args, **kwargs):
```
what happens here if `prefix` is `None`?

eddyashton replied:
This works, producing a method like `"None/getCommit"`, which will produce an error later. I've replaced the default prefixes from `None` to `"users"` to mitigate this.

cimetrics commented:
![images](https://raw.githubusercontent.com/microsoft/CCF/cimetrics/cimetrics/image2019-11-04%2012%3A51%3A20.917399.png)

cimetrics commented:
![images](https://raw.githubusercontent.com/microsoft/CCF/cimetrics/cimetrics/image2019-11-04%2013%3A05%3A28.261984.png)

cimetrics commented:
![images](https://raw.githubusercontent.com/microsoft/CCF/cimetrics/cimetrics/image2019-11-04%2013%3A30%3A29.862856.png)

cimetrics commented:
![images](https://raw.githubusercontent.com/microsoft/CCF/cimetrics/cimetrics/image2019-11-04%2014%3A12%3A58.647131.png)

eddyashton commented:
> Sorry I have to be that guy: we need to update the docs accordingly.
> At least: https://microsoft.github.io/CCF/developers/example.html and https://microsoft.github.io/CCF/users/issue_commands.html
> 
> And also update the RPC API spec?

I've updated a few places in the docs that show direct JSON queries, and added a brief description in the `issue_commands` doc.

cimetrics commented:
![images](https://raw.githubusercontent.com/microsoft/CCF/cimetrics/cimetrics/image2019-11-04%2014%3A52%3A56.982972.png)
